magma/circuit.py

```diff
@@ -830,3 +830,52 @@ def wrapped(*args, **kwargs):
 def DeclareCoreirCircuit(*args, **kwargs):
     return DeclareCircuit(*args, **kwargs,
                           renamed_ports=coreir_port_mapping)
+
+
+class _CircuitBuilderMeta(type):
+    pass
+
+
+def builder_method(func):
+
+    @wraps(func)
+    def _wrapped(this, *args, **kwargs):
+        with _DefinitionContextManager(this._context):
+            result = func(this, *args, **kwargs)
+        return result
+
+    return _wrapped
+
+
+class CircuitBuilder(metaclass=_CircuitBuilderMeta):
+    def __init__(self, name):
+        try:
+            context = _definition_context_stack.peek()
+            context.add_builder(self)
+        except IndexError:  # builders must be inside a definition context
+            raise Exception("Can not instance a circuit builder outside a "
+                            "definition")
```

Comment:
Comment is redundant with exception string

Reply:
Done.